README.md
Kafka REST Proxy
================

The Kafka REST Proxy provides a RESTful interface to a Kafka cluster. It makes it easy to produce and consume messages,
view the state of the cluster, and perform administrative actions without using the native Kafka protocol or clients.
Examples of use cases include reporting data to Kafka from any frontend app built in any language, ingesting messages
into a stream processing framework that doesn't yet support Kafka, and scripting administrative actions.

Installation
------------

You can download prebuilt versions of the Kafka REST Proxy as part of the
[Confluent Platform](http://confluent.io/downloads/). To install from source,
follow the instructions in the Development section.

Deployment
----------

The REST proxy includes a built-in Jetty server. Assuming you've configured your classpath correctly, you can start a
server with

    java io.confluent.kafkarest.Main [server.properties]

where `server.properties` contains configuration settings as specified by the `KafkaRestConfiguration` class. Although the
properties file is not required, the default configuration is not intended for production. Production deployments *should*
specify a properties file. By default the server starts bound to port 8082, does not specify a unique instance ID (required
to safely run multiple proxies concurrently), and expects Zookeeper to be available at `localhost:2181` and a Kafka broker
at `localhost:9092`.

Development
-----------

To build a development version, you may need a development version of
[io.confluent.common](https://github.com/confluentinc/common) and [io.confluent.rest-utils](https://github.com/confluentinc/rest-utils).
After installing `common` and `rest-utils` and compiling `kafka-rest` with
Maven, you can run an instance of the proxy against a local Kafka cluster (using
the default configuration included with Kafka):

    mvn exec:java

Contribute
----------

- Source Code: https://github.com/confluentinc/kafka-rest
- Issue Tracker: https://github.com/confluentinc/kafka-rest/issues

License
-------

The project is licensed under the Apache 2 license.


debian/changelog
confluent-kafka-rest (0.1~SNAPSHOT-1) unstable; urgency=low

  * Initial release.

 -- Confluent Packaging <packages@confluent.io>  Fri, 02 Dec 2015 22:09:36 +0000


src/main/java/io/confluent/kafkarest/Config.java
/**
 * Copyright 2014 Confluent Inc.
 *
 * Licensed under the Apache License, Version 2.0 (the "License");
 * you may not use this file except in compliance with the License.
 * You may obtain a copy of the License at
 *
 * http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing, software
 * distributed under the License is distributed on an "AS IS" BASIS,
 * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
 * See the License for the specific language governing permissions and
 * limitations under the License.
 */
package io.confluent.kafkarest;

import java.util.Properties;

/**
 * Settings for the REST proxy server.
 */
public class Config {
    public Time time;

    /**
     * Unique ID for this REST server instance. This is used in generating unique IDs for consumers that do not specify
     * their ID. The ID is empty by default, which makes a single server setup easier to get up and running, but is not
     * safe for multi-server deployments where automatic consumer IDs are used.
     */
    public String id;
    public static final String DEFAULT_ID = "";

    public boolean debug;
    public static final String DEFAULT_DEBUG = "false";

    public int port;
    public static final String DEFAULT_PORT = "8080";

    public String zookeeperConnect;
    public static final String DEFAULT_ZOOKEEPER_CONNECT = "localhost:2181";

    public String bootstrapServers;
    public static final String DEFAULT_BOOTSTRAP_SERVERS = "localhost:9092";

    public int producerThreads;
    public static final String DEFAULT_PRODUCER_THREADS = "5";

    /**
     * The consumer timeout used to limit consumer iterator operations. This should be very small so we can effectively
     * peek() on the iterator.
     */
    public int consumerIteratorTimeoutMs;
    public static final String DEFAULT_CONSUMER_ITERATOR_TIMEOUT_MS = "1";

    /**
     * Amount of time to backoff when an iterator runs out of data. When a consumer has a dedicated worker thread, this
     * is effectively the maximum error for the entire request timeout. It should be small enough to get reasonably close
     * to the timeout, but large enough to avoid busy waiting.
     */
    public int consumerIteratorBackoffMs;
    public static final String DEFAULT_CONSUMER_ITERATOR_BACKOFF_MS = "50";

    /** The maximum total time to wait for messages for a request if the maximum number of messages has not yet been reached. */
    public int consumerRequestTimeoutMs;
    public static final String DEFAULT_CONSUMER_REQUEST_TIMEOUT_MS = "1000";

    /** The maximum number of messages returned in a single request. */
    public int consumerRequestMaxMessages;
    public static final String DEFAULT_CONSUMER_REQUEST_MAX_MESSAGES = "100";

    public int consumerThreads;
    public static final String DEFAULT_CONSUMER_THREADS = "1";

    /** Amount of idle time before a consumer instance is automatically destroyed. */
    public int consumerInstanceTimeoutMs;
    public static final String DEFAULT_CONSUMER_INSTANCE_TIMEOUT_MS = "300000";

    public Config() throws ConfigurationException {
        this(new Properties());
    }

    public Config(Properties props) throws ConfigurationException {
        time = new SystemTime();

        id = props.getProperty("id", DEFAULT_ID);

        debug = Boolean.parseBoolean(props.getProperty("debug", DEFAULT_DEBUG));

        port = Integer.parseInt(props.getProperty("port", DEFAULT_PORT));
        zookeeperConnect = props.getProperty("zookeeper.connect", DEFAULT_ZOOKEEPER_CONNECT);
        bootstrapServers = props.getProperty("bootstrap.servers", DEFAULT_BOOTSTRAP_SERVERS);
        producerThreads = Integer.parseInt(props.getProperty("producer.threads", DEFAULT_PRODUCER_THREADS));
        consumerIteratorTimeoutMs = Integer.parseInt(props.getProperty("consumer.iterator.timeout.ms", DEFAULT_CONSUMER_ITERATOR_TIMEOUT_MS));
        consumerIteratorBackoffMs = Integer.parseInt(props.getProperty("consumer.iterator.backoff.ms", DEFAULT_CONSUMER_ITERATOR_BACKOFF_MS));
        consumerRequestTimeoutMs = Integer.parseInt(props.getProperty("consumer.request.timeout.ms", DEFAULT_CONSUMER_REQUEST_TIMEOUT_MS));
        consumerRequestMaxMessages = Integer.parseInt(props.getProperty("consumer.request.max.messages", DEFAULT_CONSUMER_REQUEST_MAX_MESSAGES));
        consumerThreads = Integer.parseInt(props.getProperty("consumer.threads", DEFAULT_CONSUMER_THREADS));
        consumerInstanceTimeoutMs = Integer.parseInt(props.getProperty("consumer.instance.timeout.ms", DEFAULT_CONSUMER_INSTANCE_TIMEOUT_MS));
    }
}
